service/sharddistributor/handler/executor_test.go

```diff
@@ -606,3 +610,241 @@ func newMigrationConfig(t *testing.T, configEntries []configEntry) *config.Migra
 	migrationConfig := config.NewMigrationConfig(dc)
 	return migrationConfig
 }
+
+func TestFilterNewlyAssignedShardIDs(t *testing.T) {
+	type testCase struct {
+		name     string
+		previous *store.HeartbeatState
+		assigned *store.AssignedState
+		expected []string
+	}
+	tests := []testCase{
+		{
+			name:     "nil previousHeartbeat returns all assigned",
+			previous: nil,
+			assigned: &store.AssignedState{
+				AssignedShards: map[string]*types.ShardAssignment{
+					"shard1": {},
+					"shard2": {},
+				},
+			},
+			expected: []string{"shard1", "shard2"},
+		},
+		{
+			name: "no new assigned shards",
+			previous: &store.HeartbeatState{
+				ReportedShards: map[string]*types.ShardStatusReport{
+					"shard1": {},
+					"shard2": {},
+				},
+			},
+			assigned: &store.AssignedState{
+				AssignedShards: map[string]*types.ShardAssignment{
+					"shard1": {},
+					"shard2": {},
+				},
+			},
+			expected: []string{},
+		},
+		{
+			name: "some new assigned shards",
+			previous: &store.HeartbeatState{
+				ReportedShards: map[string]*types.ShardStatusReport{
+					"shard1": {},
+				},
+			},
+			assigned: &store.AssignedState{
+				AssignedShards: map[string]*types.ShardAssignment{
+					"shard1": {},
+					"shard2": {},
+					"shard3": {},
+				},
+			},
+			expected: []string{"shard2", "shard3"},
+		},
+		{
+			name: "empty assigned returns empty",
+			previous: &store.HeartbeatState{
+				ReportedShards: map[string]*types.ShardStatusReport{
+					"shard1": {},
+				},
+			},
+			assigned: &store.AssignedState{
+				AssignedShards: map[string]*types.ShardAssignment{},
+			},
+			expected: []string{},
+		},
+		{
+			name: "nil assignedState returns nil",
+			previous: &store.HeartbeatState{
+				ReportedShards: map[string]*types.ShardStatusReport{
+					"shard1": {},
+				},
+			},
+			assigned: nil,
+			expected: nil,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			result := filterNewlyAssignedShardIDs(tt.previous, tt.assigned)
+			require.ElementsMatch(t, tt.expected, result)
+		})
+	}
+}
+
+func TestEmitShardAssignmentMetrics(t *testing.T) {
+	heartbeatTime := time.Now().UTC()
+	namespace := "test-namespace"
+	shardID := "shard-1"
+
+	for name, tc := range map[string]struct {
+		previousHeartbeat *store.HeartbeatState
+		assignedState     *store.AssignedState
+
+		expectedDistributionLatency *time.Duration
+		expectedHandoverLatencies   []*time.Duration
+	}{
+		"no new assigned shards": {
+			previousHeartbeat:           &store.HeartbeatState{ReportedShards: map[string]*types.ShardStatusReport{shardID: {}}},
+			assignedState:               &store.AssignedState{AssignedShards: map[string]*types.ShardAssignment{shardID: {}}},
+			expectedDistributionLatency: nil,
+			expectedHandoverLatencies:   nil,
+		},
+		"newly assigned shard with handover stats": {
+			previousHeartbeat: &store.HeartbeatState{ReportedShards: map[string]*types.ShardStatusReport{}},
```

Comment:
The `previousHeartbeat` is the same in every test. I think we should have a fixed value in the test code - we already tested the filter function in the other test